test/middleware_stack_test.rb

```diff
@@ -35,8 +35,9 @@ def test_allows_rebuilding
   end
 
   def test_allows_extending
-    build_stack Apple
+    build_handlers_stack Apple
     @conn.use Orange
+    @conn.adapter :test, &test_adapter
     assert_handlers %w[Apple Orange]
```

Comment:
Ah you're right, that makes it a little more difficult.
ActiveSupport comes with this handy extension:

```ruby
class Array
  # Extracts options from a set of arguments. Removes and returns the last
  # element in the array if it's a hash, otherwise returns a blank hash.
  #
  #   def options(*args)
  #     args.extract_options!
  #   end
  #
  #   options(1, 2)        # => {}
  #   options(1, 2, a: :b) # => {:a=>:b}
  def extract_options!
    if last.is_a?(Hash) && last.extractable_options?
      pop
    else
      {}
    end
  end
end
```

but you'll probably need to add it as it's not one of the gem dependencies.
At this point then having another method like you did is probably better, just check if you can reduce the duplication even further